[CNSMR-2053] Add Architecture.md (#194)
Babylonpartners/ios-playbook: Cookbook/Technical-Documents/Architecture.md

Luckily, web-development community has reached a more consistent architectural solution to use [Redux](https://redux.js.org) as a predictable (global) state container alongside [React](https://reactjs.org/), a state (or virtual-view) diffing renderer.
Redux encourages the developers to apply uni-directional flow architecture (inspired by [Elm](https://elm-lang.org/)) and centralize state management (powered by [Automata Theory](https://en.wikipedia.org/wiki/Automata_theory)) to solve our questions 1 and 2.
Moreover, React brings a new UI paradigm that can be rendered just from a state via Redux.

So, in Babylon iOS team, we created [ReactiveFeedback](https://github.com/Babylonpartners/ReactiveFeedback) and [Bento](https://github.com/Babylonpartners/Bento) as their counterparts to solve 1 and 2 that nicely fits on top of existing UIKit framework.

Also, for 3 and 4 (navigation), we defined `Builder` and `FlowController` layers to satisfy the remaining missing pieces.

So, in a nutshell:

1. `ViewModel` (with ReactiveFeedback)
2. `Renderer` (Bento Virtual-View)
3. `Builder`
4. `FlowController`

> are the 4 key interfaces to achieve Babylon iOS Architecture, as well as:

Reviewer: Not sure if `interfaces` are the best naming here.

```suggestion
are the 4 core entities to achieve Babylon iOS Architecture, as well as:
```

> - `Bento.BoxViewController` (universal class for rendering `UITableView`)

Author: I think `BoxViewController` is the most commonly used ViewController class to explain about our architecture.
Though we might be tempted to introduce its variety, it is not very important in this documentation since this topic is about brief overview for how each layer interacts rather than in-depth discussion of each layer.

that coordinates the real presentation in UIKit.

> 

Reviewer: Maybe we should name this architecture to make more concrete? We have mentioned MVVM-FC a few times.

Reviewer: Wondering if we should include a B for the Builder tho.

Author: The goal of this guideline is 

> Describe how VMs, VC, Renderer (Bento), FC and Builders interact

so I think `Builder` plays an important role to be included in the core entity.

I just called "Babylon iOS Architecture" in the above line, but can be changed to `MVVM + FlowController + Builder` architecture.

## Interfaces / Types

### `Builder`

**`Builder` is a screen's entrypoint** that creates `ViewModel`, `Renderer`, `FlowController`, all necessary business logics, and finally creates `ViewController` that collaborates with `ViewModel` and `Renderer` to update the actual `UITableView` / `UICollectionView` with the minimal changes via diffing algorithm by Bento.

Overall structure looks like this:

```
struct Builder

> ├── struct ViewModel: Bento.BoxViewModel

Reviewer 2: `class`

Author: ```suggestion
├── class ViewModel: Bento.BoxViewModel
```

│  ├── ViewStyleSheet<UIView>
│  └── render: (State) -> Screen<SectionID, ItemID> // creates Virtual View
│     └── Bento.Box             // Virtual View
│
├── struct FlowController
│  └── handle: (Route) -> Void  // shows next screen / presentation
│
└─ class BoxViewController      // (universal, no screen-dependent subclassing)
   └── render: (Screen) -> Void // updates actual UITableView
```

As we see in the next section, `Builder` is responsible for creating `ViewController` with `ViewController (VC) → ViewModel (VM) → FlowController (FC)` retain graph.

And `Builder` is usually created from previous screen's `(Previous)FlowController`.

### `FlowController`

**`FlowController` is the coordinator of current screen and next presentation**.

It inherits `Flow` classes from previous screen which is an abstraction of UI flow (navigation, modal, etc) to present or dismiss the screen that `Builder` is going to create.

For example, let's say "Screen0"'s `Screen0FlowController` will create a next "Screen1" using `Screen1Builder`, which then shows "Screen2" and so on:

```swift
struct Screen0FlowController {
    ...
    func handle(_ route: Route) {
        switch route {

>         case ...:

Author: Updated in 1323e27.

P.S.
- I personally am not fan of using `|>` for side-effectful function application, and it requires explanation in the doc, so I will omit it.
- I believe `builders.screen1.make()` is an old style nesting that we don't want to introduce here.

            let screen1ViewController = Screen1Builder(...).make(...)
            navigationFlow.present(screen1ViewController) // accesses `UINavigationController` to push "Screen1"
        }
    }
}

...

struct Screen1Builder {
    ...

    func make(
        navigation: Flow,
        modal: Flow,
        presenting: Flow,
        root: Flow
    ) -> UIViewController {
        // Create ViewModel.
        let viewModel = Screen1ViewModel(
            businessController: dependencies.businessController,

>             analytics: dependencies.analytics

Author: To reduce explanation of other context knowledge e.g. `Current` owning `analytics`, I will omit this change.

public protocol BoxRenderer {
    typealias Sink<Action> = (Action) -> Void
    associatedtype State
    associatedtype Action
    associatedtype SectionID: Hashable
    associatedtype ItemID: Hashable

    // `observer` as `BoxRenderer` to `BoxViewModel` event flow (user action from UI)
    init(observer: @escaping Sink<Action>, ....)

    func render(state: State) -> Screen<SectionID, ItemID>
}
```

#### Unidirectional event flow

1. `BoxRenderer.observer` is triggered with a new event
2. `BoxViewModel.send(action:)` is invoked and changes state
3. `state: Property<State>` emits a new state
4. `Bento.BoxViewController` receives a new state, creating virtual view via `BoxRenderer.render(state:)` and apply diffing to real `UITableView`
5. `UITableView` gets reloaded

### `ViewModel`

The minimal `ViewModel` example looks as follows:

```swift
class ViewModel: Bento.BoxViewModel {
    let state: Property<State>
    let routes: Signal<Route, NoError>

    private let eventPipe = Feedback<Event, NoError>.input()

    init(...) {
        self.state = Property( // Defined in ReactiveFeedback
            initial: .initial,
            reduce: ViewModel.reduce,
            feedbacks: [

>                 actionsPipe.feedback,

Reviewer 3: ```suggestion
                eventPipe.feedback,
```

                ViewModel.whenSubmitting(...)
            ]
        )

        self.routes = self.state
            .filterMap { $0.route } // State to Route conversion
    }

    // User Action from UI
    func send(action: Action) {
        eventPipe.observer(value: .ui(action))
    }

    // MARK: - Reducer

    static func reduce(_ state: State, _ event: Event) -> State {
        switch (oldState, event) {
        case ...:
            return newState
        }
    }

    // MARK: - Feedback

    static func whenSubmitting(...) {
        return Feedback(
            predicate: { $0.status.isSubmitting },
            effects: { state -> SignalProducer<Event, NoError> in
                return self.validate(state: state)
                    .flatMap(.merge) { data in

>                         self.fetchNextScreen(data)

Reviewer 3: `fetchNextScreen` is a little confusing. Are we fetching new screen? As an example I'd use something more data-specific, e.g.
```suggestion
                        self.fetchDetails(with: validatedData)
```

Author: Updated in 1323e27.

                    }

>                     .materializeResults()

Reviewer 2: I think we do not need that

Author: Same as https://github.com/Babylonpartners/ios-playbook/pull/194#discussion_r307704044 .

>                     .map(Event.didFinishSubmit)

Reviewer 2: should probably have `replaceError`

Author: It depends on how we define `Event`, e.g. in this case, 
`case didFinishSubmit(Result<NextScreenData, NetworkFailure>)` is defined.

            }
        )
    }
}
```

> - Input: User actions from `Renderer` will be delivered to `func send(action: Action)`, and forwarded as ReactiveFeedback's input.

Reviewer 3: Should we also mention other sources of input? For example API responses, permission prompt responses, etc.

Author: This PR focuses on `Describe how VMs, VC, Renderer (Bento), FC and Builders interact` , so I think other external effects don't have to be in this document for simplicity.

    - State-driven output is so-called [Moore machine](https://en.wikipedia.org/wiki/Moore_machine)

See [Babylonpartners/ReactiveFeedback](https://github.com/Babylonpartners/ReactiveFeedback) for more examples.

### Renderer

The minimal `Renderer` example looks as follows:

```swift
struct Renderer: BoxRenderer {
    private let observer: Sink<Action>
    ...

    func render(state: State) -> Screen<SectionId, RowId> {
        return Screen(
            title: "Title",
            box: .empty
                |-+ renderSection0(state: state)
                |-+ renderSection1(state: state)
                |-? renderSection2(state: state)
                |-? renderCurrentGP(state: state)
        )
    }

    private func renderSection0(state: State) -> Section<SectionId, RowId> {
        return Section<SectionId, RowId>(id: .section0)
            |---+ Node(
                id: .row0_0,
                component: InputField(
                    title: state.title,
                    placeholder: state.placeholder,
                    image: state.image,
                    didTapImage: { observer(.didTapImage) },

>                     accessibilityIdentifier: String(describing: id)

Reviewer 3: Since this is gonna be a reference for implementation I'd use `AccessibilityContent` here or something made up like `Accessibility.foo` or even `...`.
```suggestion
                    accessibilityIdentifier: ...
```
